flink-end-to-end-tests/flink-datastream-allround-test/src/main/java/org/apache/flink/streaming/tests/SequenceGeneratorSource.java

```diff
@@ -92,42 +92,63 @@
 	@Override
 	public void run(SourceContext<Event> ctx) throws Exception {
 
-		Random random = new Random();
+		if (keyRanges.size() > 0) {
```

Comment:
One suggestion because this methods became a bit long, why not break it down a bit? I would introduce two private methods, e.g., `runActive()` and `runIdle()` which are called in the `if` branches.